corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/api.js

```diff
@@ -171,7 +175,22 @@ hqDefine("cloudcare/js/formplayer/menus/api", function () {
                 });
 
                 var callStartTime = performance.now();
-                menus.fetch($.extend(true, {}, options)).always(function () {
+                console.log(`currentSelection:  ${ JSON.stringify(currentSelections) }, new selections: ${ JSON.stringify(params.selections) }`);
+                if (!_.isEqual(params.selections, currentSelections)) {
+                    console.log('different selection: interrupt ongoing defers')
+                    currentSelections = params.selections;
+                    while (ongoingRequests.length > 0) {
+                        const ongoingRequest = ongoingRequests.pop();
+                        if (ongoingRequest.readyState !== 4) {
+                            console.log('found pending defer. rejecting it.')
+                            ongoingRequest.abort();
+                        }
+                    }
+                }
+                const updateRequest = menus.fetch($.extend(true, {}, options));
+                ongoingRequests.push(updateRequest);
```

Comment:
This feels like a potentially broad-reaching change - what kinds of requests will pop up here?  Is there a chance of any weirdness with certain request types being interrupted?  What are the advantages of this approach versus say a locking approach (preventing new requests until the old one is finished)?

Reply:
I suppose all form player requests go through here (navigate_menu, get_endpoint, get_details from a quick look). I could limit this behavior to just the navigate_menu requests since those are the target.

The bug was found when somebody tried to click in the bread crumb bar while the search results were loading. Since that can take a long time it could make the UI appear unresponsive if we went with the waiting to finish route.

Reply:
@AddisonDunn The array in persisted in the js environment which should be just a single tab (99% sure about that). The selections are stored in the session storage which was previously updated out of order hence this fix. But that as well is only within a tab. Otherwise you would not be able to open the same web app twice and and navigate to different forms.

Reply:
Food for thought on @esoergel 's original question: have you read task_queue.js and seen [how it's used in web_form_session](https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/cloudcare/static/cloudcare/js/form_entry/web_form_session.js#L103-L108) to prevent multiple form submission requests?

This isn't quite the same use case, and I'm not asking for changes on this PR, just sharing it as an example of encapsulating somewhat similar behavior.